var/spack/repos/builtin/packages/faiss/package.py

```diff
@@ -52,39 +65,225 @@ class Faiss(AutotoolsPackage, CudaPackage):
+    #   cpu_targets and gpu_targets as two lists
+    #   cpu_installs and gpu_installs as two dictionaries
+    #       where key = build subdir, and val = list of exe to copy
+    def fetch_targets_v163(self):
+
+        ctargets = ["tests"]
+        gtargets = ["build", "demo_ivfpq_indexing_gpu"]  # target added by the patch
+
+        cinstalls = {"tests": ["tests"]}
+        ginstalls = {
+            "gpu/test": [
+                "TestGpuIndexFlat",
+                "TestGpuIndexBinaryFlat",
+                "TestGpuIndexIVFFlat",
+                "TestGpuIndexIVFPQ",
+                "TestGpuMemoryException",
+                "TestGpuSelect",
+                "demo_ivfpq_indexing_gpu",
+            ]
+        }
+
+        return (ctargets, gtargets), (cinstalls, ginstalls)
+
+    def fetch_targets_v172(self):
+
+        ctargets = [
+            "demo_imi_flat",
+            "demo_imi_pq",
+            "demo_ivfpq_indexing",
+            "demo_nndescent",
+            "demo_sift1M",
+            "demo_weighted_kmeans",
+        ]
+
+        gtargets = [
+            "TestCodePacking",
+            "TestGpuDistance",
+            "TestGpuIndexBinaryFlat",
+            "TestGpuIndexFlat",
+            "TestGpuIndexIVFFlat",
+            "TestGpuIndexIVFPQ",
+            "TestGpuIndexIVFScalarQuantizer",
+            "TestGpuSelect",
+            "TestGpuMemoryException",
+            "demo_ivfpq_indexing_gpu",
+        ]
+
+        cinstalls = {"tests": ["faiss_test"], "demos": ctargets}
+        ginstalls = {"faiss/gpu/test": gtargets}
+
+        return (ctargets, gtargets), (cinstalls, ginstalls)
+
+    # --------------------------------------------------------------------------
+    @when("@1.7.2:")
+    def cmake_args(self):
+
+        variants2cmake = {
+            "cuda": "FAISS_ENABLE_GPU",
+            "python": "FAISS_ENABLE_PYTHON",
+            "tests": "BUILD_TESTING",
+            "shlib": "BUILD_SHARED_LIBS",
+        }
+
+        cmake_args = []
+        for v, c in variants2cmake.items():
+            if "+" + v in self.spec:
+                cmake_args.append("-D" + c + "=ON")
+            else:
+                cmake_args.append("-D" + c + "=OFF")
+
+        return cmake_args
+
+    # --------------------------------------------------------------------------
+    def build(self, spec, prefix):
+
+        with working_dir(self.build_directory):
+
+            # make the faiss build
+            make("faiss")
+
+            # build python bindings
+            if "+python" in self.spec:
+                make("swigfaiss")
+
+            if "+tests" not in self.spec:
+                return
+
+            # build tests
+            (ctargets, gtargets), (_, _) = self.fetch_targets_v172()
+
+            # CPU tests
+            make("gtest")
+            for t in ctargets:
+                make(t)
+
+            # GPU tests
+            if "+cuda" in self.spec:
+                for t in gtargets:
+                    make(t)
```

Comment:
I think you don't need to override the phase. It's possibly feasible to just define the `build_targets` property. For tests you can override the `check` method.

Reply:
you're right. I tried that and it works fine. But for older versions, the build_directory is different too and targets etc are different. I was not able to get the "build_targets" condition on the version, so I am leaving this as is for now.

Tests need to be buile and installed too.

my experience with faiss has been that they keep changing their build quite substantially. while i agree that we could utilize several functionality to keep the package.py file simple here, I am seeing additional complexity and conditional functions that I believe offset the gains. perhaps it's easier to keep the older version build written as is, and the newer versions we can continue to improve